### PR TITLE
Support different lshape maps in binary ops

### PR DESCRIPTION
**Pull request by ClaudiaComito**
## Description

Implementing support for binary operations on distributed DNDarrays  of same shape, same split, but different distribution map, i.e.:
```python
a =  ht.ones(10, split=0)
b = ht.zeros(10, split=0)
c = a[:-1] + b[1:]
```

Issue/s resolved: #880 

## Changes proposed:
- `redistribute_` one of the two DNDarrays to match the distribution map of the other.
- expand tests
- include a warning in the `__binary_op` documentation

## Type of change

- New feature (non-breaking change which adds functionality)

## Memory requirements

Memory requirements are consistent with the extra memory requirements of `DNDarray.redistribute_()`.

## Performance

Performance slow-down is consistent with redistributing a DNDarray before a binary operation.

## Due Diligence
- [x] All split configurations tested
- [x] Multiple dtypes tested in relevant functions
- [x] Documentation updated (if needed)
- [x] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no, but it redistributes one of the operands in place.

### CODE REVIEW

File: heat/core/_operations.py [hunk 1]
```diff
@@ -119,29 +154,37 @@ def __binary_op(
...
-                    color = 0 if t1.comm.rank < t2.shape[t1.split] else 1
-                    newcomm = t1.comm.Split(color, t1.comm.rank)
-                    if t1.comm.rank > 0 and color == 0:
-                        t1.larray = torch.zeros(
-                            t1.shape, dtype=t1.dtype.torch_type(), device=t1.device.torch_device
-                        )
-                    newcomm.Bcast(t1)
-                    newcomm.Free()
-
+                    t1.resplit_(None)
```
Comment (ben-bou) [inline heat/core/_operations.py:157]: Why is this needed? It should work the same without this resplit, no?
Nothing below depends on having at least two elements in the split dimension

File: heat/core/_operations.py [hunk 2]
```diff
@@ -153,23 +196,43 @@ def __binary_op(
...
+        # TODO: implement `dndarray.create_bulk_lshape_maps`
+        t1.create_lshape_map()
+        t2.create_lshape_map()
         if len(t1.lshape) > t1.split and t1.lshape[t1.split] == 0:
             result = t1.larray.type(promoted_type)
         else:
+            if (
+                t2.split is not None
+                and not (t2.lshape_map[:, t2.split] == t1.lshape_map[:, t1.split]).all()
+                and t2.lshape_map[:, t2.split].sum() == t1.lshape_map[:, t1.split].sum()
```
Comment (ben-bou) [inline heat/core/_operations.py:210]: The last condition is not necessary:
1) `output_shape` is already determined, so shapes are compatible
2) no (shape)broadcasting is happening in the split dimension because of lines 152 to 175
-> elements in split-dimension are already equal

**Comment by ClaudiaComito** [comment]
Changes addressed in #902 by @ben-bou
